crates/app/tests/mod.rs

```diff
@@ -5,45 +5,50 @@ mod helpers;
 
 use helpers::*;
 
-use std::net::{Ipv4Addr, TcpListener};
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
 use std::time::Duration;
 
 use drawbridge_app::Builder;
+use drawbridge_test_helper::TestHelper;
 use drawbridge_type::digest::Algorithms;
 use drawbridge_type::{RepositoryConfig, TagEntry, TreeEntry};
 
-use axum::Server;
-use futures::channel::oneshot::channel;
-use reqwest::StatusCode;
+use reqwest::{Client, StatusCode};
 use tempfile::tempdir;
 
-#[tokio::test]
-async fn app() {
-    let lis = TcpListener::bind((Ipv4Addr::UNSPECIFIED, 0)).unwrap();
-    let addr = format!("http://{}", lis.local_addr().unwrap());
+async fn app_tests(th: &TestHelper, cl: Client) {
+    let lis = SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 8080);
+    let addr = format!("https://{}", lis);
     let store = tempdir().expect("failed to create temporary store directory");
 
-    let (tx, rx) = channel::<()>();
-    let srv = tokio::spawn(
-        Server::from_tcp(lis)
-            .unwrap()
-            .serve(Builder::new(store.path()).build().unwrap())
-            .with_graceful_shutdown(async { rx.await.ok().unwrap() }),
-    );
-
-    let cl = reqwest::Client::builder()
-        .timeout(Duration::new(1, 0))
-        .build()
-        .unwrap();
-
+    let graceful_shutdown = Arc::new(AtomicBool::new(false));
+    let graceful_shutdown_clone = graceful_shutdown.clone();
+    let pem_certificates = th.server_pem.clone();
+    let pem_private_key = th.server_private_pem.clone();
+    let srv = tokio::spawn(async move {
+        Builder::new(store.path())
+            .serve(
+                &lis,
+                &pem_certificates,
+                &pem_private_key,
+                // th.ca_pem,
```

Comment:
passing pem_ca I get this

```
thread 'app' panicked at 'called `Result::unwrap()` on an `Err` value: reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(8080), path: "/", query: None, fragment: None }, source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })) }', crates/app/tests/mod.rs:46:42
```
which is strange because going back to passing None works again

Reply:
```rust
    let aaa = th.ca_pem.clone();
    let srv = tokio::spawn(async move {
        Builder::new(store.path())
            .serve(
                &lis,
                &pem_certificates,
                &pem_private_key,
                Some(aaa),
                // None,
                graceful_shutdown_clone,
            )
            .await
            .map_err(|e| format!("{}", e))
    });
    // Wait for the server to start
    tokio::time::sleep(Duration::from_secs(1)).await;
    ```